### PR TITLE
fix: filters in chroma integration

### LINKED ISSUE #1010: Using multiple arguments for FilterRetriever
**Describe the bug**
Hi, I am using the FilterRetriever and having problems with applying more than one filter.

This from a Chroma vectorstore. WITH ONE FILTER it works, I can not fiugre out how to put multiple filters.

the pseudocode:
retriever = FilterRetriever(context_document_store)

report_filters = {
    'meta-key-a':  param-a,
    'meta-key-b'': param-b
}

__this does not work:__
result = retriever.run(report_filters)

__this does work:__
result = retriever.run({ 'meta-key-a':  param-a})

In other words i can only pass one argument to the filter. Is this a problem with Chroma or whay am i missing here.

**To Reproduce**
context_document_store = ChromaDocumentStore()

document_pipeline = Pipeline()
document_pipeline.add_component("converter", MarkdownToDocument())
document_pipeline.add_component("cleaner", DocumentCleaner())
document_pipeline.add_component("splitter", DocumentSplitter(split_by="sentence", split_length=10))
document_pipeline.add_component("writer", DocumentWriter(document_store=context_document_store))
document_pipeline.connect("converter", "cleaner")
document_pipeline.connect("cleaner", "splitter")
document_pipeline.connect("splitter", "writer")

document_pipeline.run({"converter": {
    "sources": ["vaud_report_results.md", "report_results.md", "lac_leman_g70.md"], 
    'meta':[
        {'doc-id': 'Vaud 2015-11-15 2021-12-31', 'topic':'reports'}, 
        {'doc-id': 'Bern 2015-11-15 2021-12-31', 'topic':'reports'},
        {'doc-id': 'Lac Léman 2015-11-15 2021-12-31', 'topic': 'reports'}
    ]
    }})

pipeline = Pipeline()
pipeline.add_component("converter", PyPDFToDocument())
pipeline.add_component("cleaner", DocumentCleaner())
pipeline.add_component("splitter", DocumentSplitter(split_by="page", split_length=1))
pipeline.add_component("writer", DocumentWriter(document_store=context_document_store))
pipeline.connect("converter", "cleaner")
pipeline.connect("cleaner", "splitter")
pipeline.connect("splitter", "writer")

file_names = [
    'resources/brief_history_marine_litter.pdf', 
    'resources/coastline_litter_threshold_value_report_14_9_2020_final.pdf', 
    'resources/revealing_the_role_of_landuse.pdf',
    'resources/Walvoort-ea-2021-Modelling-Forecasting-Beach-Litter-Assessment-Values-1.pdf',
    'resources/eu-guide-marine-litter-2023.pdf',
    'resources/land-use-marine-litter-malaysia.pdf'
]
metas = [
    {'topic': 'history of research, methods of research'},  
    {'topic': 'threshold values, methods of calculation'}, 
    {'topic': 'geospatial analysis, land use, feature evaluation'},
    {'topic': 'threshold values, methods of calculation'},
    {'topic': 'sampling protocols, methods of research'},
    {'topic': 'geospatial analysis, land use, feature evaluation'}
]

pipeline.run({"converter": {"sources": file_names, "meta": metas}})

filters = { "operator": "AND",
      			"conditions": [
        			{"field": "meta.topic", "operator": "in", "value":['threshold values, methods of calculation']},
                    {"field": "meta.doc-id", "operator": "in", "value":['Lac Léman 2015-11-15 2021-12-31']},
      					]
    			   }
retriever = FilterRetriever(context_document_store)
result = retriever.run(filters)
**Describe your environment (please complete the following information):**
 - OS: Pop!_OS 22.04
 - Haystack version: haystack-ai 2.31 (from conda list)
 - Integration version:

### PR DESCRIPTION
**Pull request by Amnah199**
### Related Issues

- fixes #1010 

### Proposed Changes:

- Fixed the logical and conversion errors in `normalize_filters` function.
- The example in docstrings were also not accurate as the chromadb filters require array inputs for logical operators ([reference docs)](https://cookbook.chromadb.dev/core/filters/). I updated the examples to follow the chroma db docs.

### How did you test it?

-  added test for checking nested operators in filters

### Notes for the reviewer

If you have any particular example to test with filters, you can notify me to add them in tests.
### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.

### CODE REVIEW

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py [hunk 1]
```diff
@@ -363,50 +350,47 @@ def _normalize_filters(filters: Dict[str, Any]) -> Tuple[List[str], Dict[str, An
...
-                # if the list contains multiple items, we need an $or chain
-                for v in value:
-                    where["$or"].append({field: v})
-
-        for k in keys_to_remove:
-            del filters[k]
-
-        final_where = dict(filters)
-        final_where.update(dict(where))
+                    where[field] = value[0]  # Simplify single-item lists
```
Comment (shadeMe) [inline integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:376]: Let's provide an example in the comment to aid readability.

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py [hunk 2]
```diff
@@ -363,50 +350,47 @@ def _normalize_filters(filters: Dict[str, Any]) -> Tuple[List[str], Dict[str, An
...
+                # Check if list contains special operators that apply to documents
+                elif any(isinstance(v, dict) and any(k in v for k in ["$contains", "$not_contains"]) for v in value):
+                    where_document = filters.copy()
```
Comment (shadeMe) [inline integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:379]: Why are we copying the partially modified filters dict here? This will lead to different outputs depending on the ordering of the dict items.

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py [hunk 3]
```diff
@@ -363,50 +350,47 @@ def _normalize_filters(filters: Dict[str, Any]) -> Tuple[List[str], Dict[str, An
...
-                    where["$or"].append({field: v})
-
-        for k in keys_to_remove:
-            del filters[k]
-
-        final_where = dict(filters)
-        final_where.update(dict(where))
+                    where[field] = value[0]  # Simplify single-item lists
+                # Check if list contains special operators that apply to documents
+                elif any(isinstance(v, dict) and any(k in v for k in ["$contains", "$not_contains"]) for v in value):
```
Comment (shadeMe) [inline integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:378]: I feel like this `elif` clause can be simplified (after switching a local accumulator as mentioned earlier).

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py [hunk 4]
```diff
@@ -363,50 +350,47 @@ def _normalize_filters(filters: Dict[str, Any]) -> Tuple[List[str], Dict[str, An
...
-                validate_where(final_where)
-            if where_document:
+            if document_flag:
                 validate_where_document(where_document)
+            elif filters or where:
+                final_where.update(filters)
+                final_where.update(where)
+                validate_where(final_where)
         except ValueError as e:
             raise ChromaDocumentStoreFilterError(e) from e
```
Comment (shadeMe) [inline integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:395]: I think we can provide some extra context here, i.e., mention what operation raise the error.

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py [hunk 5]
```diff
@@ -453,7 +437,7 @@ def _query_result_to_documents(result: QueryResult) -> List[List[Document]]:
             for j in range(len(answers)):
                 document_dict: Dict[str, Any] = {
                     "id": result["ids"][i][j],
-                    "content": documents[i][j],
+                    "content": answers[j],
```
Comment (shadeMe) [inline integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:440]: Why was this change made? Doesn't seem related.

**Review by shadeMe — CHANGES_REQUESTED** [review]

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py [hunk 6]
```diff
@@ -363,50 +350,47 @@ def _normalize_filters(filters: Dict[str, Any]) -> Tuple[List[str], Dict[str, An
...
-                keys_to_remove.append(field)
                 ids.append(value)
-            elif isinstance(value, (list, tuple)):
-                # Schedule for removal the original key, we're going to change it
-                keys_to_remove.append(field)
-
-                # if the list is empty the filter is invalid, let's just remove it
-                if len(value) == 0:
+                del filters[field]
+            elif isinstance(value, (list, tuple)) and field not in ["$and", "$or"]:
```
[the remark this answers is not in the document]
Comment (Amnah199) [reply integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:370]: You are right. I think its a mistake, because the chromadb filters also don't support tuples. So we basically don't have this functionality. There is no mention of them in docs so we can just remove them from here?

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py (hunk 5 above)
Comment (Amnah199) [reply integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:440]: I think it was modified while I was trying to use --fix from lint scripts. Will fix it.

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shadeMe) [reply integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:370]: Ah, then yeah - We can remove it without a deprecation (counts as a bug).

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py (hunk 5 above)
Comment (Amnah199) [reply integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:440]: Without this change it raises a linting error, which is also kinda valid.

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py (hunk 4 above)
Comment (Amnah199) [reply integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:395]: The error message generated by the internal chroma function usually highlights which part of the filter is problematic. So I think this should be sufficient.

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py (hunk 3 above)
Comment (Amnah199) [reply integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py]: I believe we need this to check if there is any document filter that needs to be treated differently.

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py (hunk 4 above)
Comment (shadeMe) [reply integrations/chroma/src/haystack_integrations/document_stores/chroma/document_store.py:399]: Sure, but we should prefix that with something along the lines of `Failed to validate WhereDocument` (unless the library mentions that too).

**Review by shadeMe — CHANGES_REQUESTED** [review]

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/filters.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    try:
+        if where_document:
+            test_clause = "document content filter"
+            validate_where_document(where_document)
+        elif where:
+            test_clause = "metadata filter"
+            validate_where(where)
+    except ValueError as e:
+        msg = f"Invalid '{test_clause}' : {e}"
+        raise ChromaDocumentStoreFilterError(msg) from e
```
Comment (Amnah199) [inline integrations/chroma/src/haystack_integrations/document_stores/chroma/filters.py:67]: It raises an error message like below:
`Invalid 'document content filter' : Expected where document operator to be one of $contains, $and, $or, got $eq`

Let me know if you think it can be improved.

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/filters.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+    return ids, where, where_document
+
+
+def _normalize_filters(filters: Dict[str, Any]) -> Dict[str, Any]:
```
Comment (shadeMe) [inline integrations/chroma/src/haystack_integrations/document_stores/chroma/filters.py:72]: Not sure if "normalize" is the right verb here - we're just converting one pre-normalized format to another. 

I think we can inline this function in `_convert_filters`:
```python
converted_clause = _parse_comparison_condition(clause) if "field" in clause else _parse_logical_condition(clause)
```

**Review by shadeMe — CHANGES_REQUESTED** [review]
Looks much better! Some more comments.

File: integrations/chroma/src/haystack_integrations/document_stores/chroma/filters.py (hunk 8 above)
Comment (Amnah199) [reply integrations/chroma/src/haystack_integrations/document_stores/chroma/filters.py:72]: Will cause a bit of repetition as this method used in line 117 of this file. Give a thumbs up if you still prefer to replace it.

Comment (shadeMe) [reply integrations/chroma/src/haystack_integrations/document_stores/chroma/filters.py:72]: Ah, didn't notice that! We can keep it then, but we need a better name - there's no actual normalization going on, just conversion. `_convert_filter_clause` or something along those lines?

**Review by shadeMe — APPROVED** [review]
Great work! 🎉
